josh-json/src/index.js

```diff
@@ -0,0 +1,524 @@
+const { get: _get, set: _set, unset, isFunction } = require('lodash');
+const Err = require('./error');
+const { FileManager } = require('./files.js');
+const uuidv4 = require('uuid').v4;
+const path = require('path');
+
+class JoshProvider {
+  constructor(options = {}) {
+    this.options = options;
+    this.dir = this.options.dataDir ?
+      path.resolve(this.options.dataDir) :
+      './data';
+    this.files = new FileManager(this.dir, options.providerOptions);
+  }
+  /**
+   * Internal method called on persistent joshs to load data from the underlying database.
+   * @param {Map} josh In order to set data to the josh, one must be provided.
+   * @returns {Promise} Returns the defer promise to await the ready state.
+   */
+  async init() {
+    return true;
+  }
+
+  /**
+   * Fetch the value of a key in the database
+   * @param {(string|number)} key The document key to fetch
+   * @param {(string|null)} path The path in the document to get
+   * @returns {Promise} Resolves to the value
+   * @example
+   * ```
+   * await JoshProvider.get("josh")
+   * ```
+   */
+  async get(key, path = null) {
+    const data = await this.files.getData(key);
+    return path ? _get(data, path) : data;
+  }
+
+  async getAll() {
+    await this.check();
+    const data = await this.files.getData();
+    return data;
+  }
+  /**
+   * Fetch multiple keys from the database
+   * @param {string[]} arr Multiple keys to fetch from the database
+   * @returns {Object} Returns object with each key mapped to its value
+   * @example
+   * ```
+   * await JoshProvider.getMany([ "josh", 123 ])
+   * ```
+   */
+  async getMany(arr) {
+    await this.check();
+    const data = await this.files.getData();
+    const final = {};
+    Object.entries(data).forEach(([key, val]) => {
+      if (arr.includes(key)) {
+        final[key] = val;
+      }
+    });
+    return final;
+  }
+
+  async random(count = 1) {
+    await this.check();
+    const docs = Object.entries(await this.getAll()) || [];
+    const final = {};
+    if (docs.length === 0) return {};
+    for (let n = 0; n < count; n++) {
+      const [key, val] = docs[Math.floor(Math.random() * docs.length)];
+      final[key] = val;
+    }
+    return final;
+  }
+
+  async randomKey(count = 1) {
+    await this.check();
+    const docs = Object.entries(await this.random(count));
+    return docs.map(([key]) => key);
+  }
+
+  /**
+   * Check for key in database
+   * @param {(string|number)} key The document key to check
+   * @param {string} path The path in document to check
+   * @return {boolean} True if key is found in database
+   * @example
+   * ```
+   * await JoshProvider.has("number")
+   * ```
+   */
+  async has(key, path = null) {
+    await this.check();
+    return await this.get(key, path) != null;
```

Comment:
It might be more performant to check whether the `indexes` contains that key, I think.

Reply:
Yup I agree

Reply:
One issue would be that path would still need to eventually pull the value